src/App.tsx

```diff
@@ -1,24 +1,162 @@
+import { Todo } from './types/Todo';
+import {
+  addTodo,
+  deleteTodo,
+  getTodos,
+  updateTodo,
+} from './api/todos';
+import { TodoFilter } from './components/TodoFilter/TodoFilter';
+import { FilterBy } from './utils/enums';
+import { TodoForm } from './components/TodoForm/TodoForm';
+import { TodoPatch } from './types/TodoPatch';
 
-const USER_ID = 0;
+const USER_ID = 10897;
 
 export const App: React.FC = () => {
-  if (!USER_ID) {
-    return <UserWarning />;
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [tempTodo, setTempTodo] = useState<null | Todo>(null);
+  const [loadingTodoIds, setLoadingTodoIds] = useState<number[]>([]);
+  const [filterBy, setFilterBy] = useState(FilterBy.All);
+  const [errorText, setErrorText] = useState<null | string>(null);
+  const [hasError, setHasError] = useState(false);
+
+  const resetError = useCallback(() => {
+    setHasError(false);
+  }, []);
+
+  const setError = (errorTxt: string) => {
+    setErrorText(errorTxt);
+    setHasError(true);
+    setTimeout(resetError, 3000);
+  };
+
+  useEffect(() => {
+    getTodos(USER_ID)
+      .then(setTodos)
+      .catch((error) => {
+        setError(error.message);
+      });
+  }, []);
+
+  const onFilterChange = useCallback((filter: FilterBy) => {
+    setFilterBy(filter);
+  }, []);
+
+  const onAddTodo = async (title: string) => {
+    if (!/\S/g.test(title)) {
+      setError('Title can\'t be empty');
+
+      return false;
+    }
+
+    const newTodo = {
+      title,
+      userId: USER_ID,
+      completed: false,
+    };
+
+    setTempTodo({ ...newTodo, id: 0 });
+
+    try {
+      const addedTodo = await addTodo(newTodo);
+
+      setTodos(prevTodos => [...prevTodos, addedTodo]);
+    } catch (error) {
+      setError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+    }
+
+    return true;
+  };
+
+  const onDeleteTodo = useCallback(async (todoId: number) => {
+    setLoadingTodoIds(prevIds => [...prevIds, todoId]);
+
+    try {
+      await deleteTodo(todoId);
+
+      setTodos(prevTodos => prevTodos.filter(todo => todo.id !== todoId));
+    } catch (error) {
+      setError('Unable to delete a todo');
+    } finally {
+      setLoadingTodoIds(prevIds => prevIds.filter(id => id !== todoId));
+    }
+  }, []);
+
+  const onUpdateTodo = useCallback(async (todoId: number, data: TodoPatch) => {
+    setLoadingTodoIds(prevIds => [...prevIds, todoId]);
+
+    try {
+      await updateTodo(todoId, data);
+
+      setTodos(prevTodos => prevTodos.map(todo => {
+        if (todo.id === todoId) {
+          return {
+            ...todo,
+            ...data,
+          };
+        }
+
+        return todo;
+      }));
+    } catch (error) {
+      setError('Unable to update a todo');
+    } finally {
+      setLoadingTodoIds(prevIds => prevIds.filter(id => id !== todoId));
+    }
+  }, []);
+
+  const filterTodos = (todosArray: Todo[], filter: FilterBy) => {
```

Comment:
This function `filterTodos` could be moved outside of the component. This would improve performance as it won't be recreated each time the component rerenders. Remember to pass all dependencies as parameters to the function.